### ISSUE TITLE
process.activateUvLoop is not called for the new node internalNextclick implementation (electron 1.8+)

### ISSUE THREAD

**Issue by xaviergonz** [electron/electron#12108]
Electron version: 1.8+
Operating system: Windows 10 x64 (maybe others)

When running the following code inside a main process:
```
var async_hooks = require('async_hooks');

let indent = 0;
async_hooks.createHook({
  init(asyncId, type, triggerAsyncId) {
    const eid = async_hooks.executionAsyncId();
    const indentStr = '-'.repeat(indent);
    console.log(
      `${indentStr}${type}(${asyncId}):` +
      ` trigger: ${triggerAsyncId} execution: ${eid}`);
  },
  before(asyncId) {
    const indentStr = '-'.repeat(indent);
    console.log(`${indentStr}before:  ${asyncId}`);
    indent += 2;
  },
  after(asyncId) {
    indent -= 2;
    const indentStr = '-'.repeat(indent);
    console.log(`${indentStr}after:   ${asyncId}`);
  },
  destroy(asyncId) {
    const indentStr = '-'.repeat(indent);
    console.log(`${indentStr}destroy: ${asyncId}`);
  },
}).enable();

require('net').createServer(() => {}).listen(8086, () => {
});
```

We get the following correct output (trimmed for clarity):
```
TCPWRAP(5): trigger: 0 execution: 0
TickObject(6): trigger: 5 execution: 0 // this is the enqueue of internal next tick that will emit the listening/error event
before:  6 // before emitting the event
after:   6 // after emitting the event
destroy: 6
```

However, when running that same code under a renderer process (for example in the console) we get only this:
```
TCPWRAP(3): trigger: 0 execution: 0
TickObject(4): trigger: 3 execution: 0
```

and it is not until we call something like process.nextTick or setImmediate that the actual enqueued TickObject executes
This is, if after that we run ```setImmediate(() => {});```
Then we resume the execution of the held queue:
```
Immediate(5): trigger: 0 execution: 0
before:  5
after:   5
before:  4 // before emitting the event
after:   4 // after emitting the event
destroy: 5
destroy: 4
```

PS: I run the same thing on a renderer with NW.js (Chromium 64 + Node 9.6.1), so the introduced bug seems not to be related to the new v8 / node version, but rather with something inside electron itself.

This is related to issue https://github.com/electron/electron/issues/11086

**Comment by xaviergonz** [comment]
Apparently calling process._tickCallback() has the same effect as calling setImmediate, shouldn't that be called by electron itself though?

**Comment by xaviergonz** [comment]
so far I'm fixing it by adding this to my renderer processes:
```
const noop = () => {
  // empty
};
setInterval(() => {
  setImmediate(noop);
}, 500);
```

But there must be a better way...

**Comment by xaviergonz** [comment]
Actually the important part is not really setImmediate but what it does internally, this is, calling process.activateUvLoop();, which seems to run pending events on the main thread, as seen in https://github.com/electron/electron/blob/7ef69a5af595c9c0eef27bdf9446103776917aec/lib/common/init.js

This worked on electron <= 1.7 because node internally used to use process.nextTick to invoke such events, and electron's implementation override that implementation to call activateUvLoop.

However the new node implementation sometimes uses an internal function instead (const { nextTick } = require('internal/process/next_tick');), which doesn't implement such workaround (see https://github.com/nodejs/node/blob/master/lib/internal/process/next_tick.js)

The only solution I can think of is to either:
a) patch node source code so it allows patching exports.nextTick (which is the internalNextTick)

b) run at certain intervals process.activateUvLoop(), which would add a delay, but only for calls to internalNextTick

c) ask node guys to expose _internalNextTick inside process, to change the calls back to process._internalNextTick (there aren't that many, just a few for http/network stuff) and then wrap that call like it is currently done for setImmediate/nextTick)

**Comment by xaviergonz** [comment]
I just created an issue for solution C in https://github.com/nodejs/node/issues/19104

**Comment by xaviergonz** [comment]
Actually our node friends just gave us an option D, which is even better. Using async_hooks instead of monkey patching to call process.activateUvLoop.

```
const asyncHooks = require('async_hooks');

  const asyncHook = asyncHooks.createHook({
    init(asyncId, type, triggerAsyncId) {
      switch (type.toUpperCase()) {
        case 'IMMEDIATE':
        case 'TIMEOUT':
        case 'TICKOBJECT':
          process.activateUvLoop();
          break;
        default:
          break;
      }
    },
  });
  asyncHook.enable();

```

that'd be required to run for renderer process.

**Comment by xaviergonz** [comment]
Just created a pull request for it.
Apparently this fix will also fix a future issue that would arise with node 10.x, where they will use setUnrefTimeout for http timeouts and similar, but since it uses the same Timeout object under the hood it triggers the same async hook.

**Comment by ckerr** [comment]
I like the idea of using existing API instead of intercepting `process.nextTick` and `global.setImmediate`, but I'll leave the review to someone else more familiar with Node's async API than me.

Though if this change does depend on an upstream fix in Node (https://github.com/nodejs/node/issues/19104, we'll have to decide whether this should be fixed in 1.8 and 2.0 with an extra patchfile for Node or whether to pick up our part of the fix after Node has a release containing their part.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by sofianguy** [comment]
Thank you for your issue!

We haven't gotten a response to our questions in our comment above. With only the information that is currently in the issue, we don't have enough information to take action. I'm going to close this but don't hesitate to reach out if you have or find the answers we need, we'll be happy to reopen the issue.
